### PR TITLE
Remove SpoonHTTP

### PR DESCRIPTION
**Pull request by tijsverkoyen**
As you all know Spoon isn't supported anymore, therefore we are trying to 
replace all functionality that was provided by Spoon with SYmfony-functionality
or native code. 

With this upgrade we have replaced all functionality provided with SpoonHTTP. 
There are several methods, each one is discussed below.

* [x] SpoonHTTP::getContent()
* [x] SpoonHTTP::getHeadersList()
* [x] SpoonHTTP::getIp()
* [x] SpoonHTTP::redirect()
* [x] SpoonHTTP::setHeaders()
* [x] SpoonHTTP::setHeadersByCode()

This will fix https://github.com/forkcms/forkcms/issues/1224

### CODE REVIEW

**Comment by tijsverkoyen** [comment]
Note to self:
`SpoonHTTP::getHeadersList()`and `SpoonHTTP::getIp()` are not used in Fork.

File: UPGRADE_3.11.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+Before:
+
+```php
+SpoonHTTP::getContent($url)
+```
+
+After:
+
+```php
+file_get_content($url)
```
Comment (WouterSioen) [inline UPGRADE_3.11.md]: typo, should be file_get_contents

File: UPGRADE_3.11.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+Before:
+
+```php
+\SpoonHTTP::redirect($url);
+```
+
+After:
+
+```php
+throw new RedirectException(
```
Comment (WouterSioen) [inline UPGRADE_3.11.md]: Maybe add the $this->redirect option too, because this is the preferred way in action files.

File: UPGRADE_3.11.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+```
+
+After:
+
+```php
+header('header');
+```
+
+But you really should think about using this, as the correct way is to bubble 
+set the headers on the response object and bubble it up.
```
Comment (WouterSioen) [inline UPGRADE_3.11.md]: typo in this sentence

Comment (tijsverkoyen) [inline UPGRADE_3.11.md]: fixed in 351f3ce9fbb9a00a67e0afb3c767c6801bceb8bc

File: UPGRADE_3.11.md (hunk 2 above)
Comment (tijsverkoyen) [inline UPGRADE_3.11.md]: fixed in 351f3ce9fbb9a00a67e0afb3c767c6801bceb8bc

File: UPGRADE_3.11.md (hunk 1 above)
Comment (tijsverkoyen) [inline UPGRADE_3.11.md]: fixed in 351f3ce9fbb9a00a67e0afb3c767c6801bceb8bc

File: src/Frontend/Modules/FormBuilder/Widgets/Form.php [hunk 4]
```diff
@@ -518,8 +520,10 @@ private function validateForm()
                 $redirect .= (stripos($redirect, '?') === false) ? '?' : '&';
                 $redirect .= 'identifier=' . $this->item['identifier'];
 
-                // redirect with identifier
-                \SpoonHTTP::redirect($redirect);
+                throw new RedirectException(
```
Comment (WouterSioen) [inline src/Frontend/Modules/FormBuilder/Widgets/Form.php]: Can't you use $this->redirect in here? Or is this not available in widgets?

Comment (tijsverkoyen) [inline src/Frontend/Modules/FormBuilder/Widgets/Form.php]: It isn't available in widgets.

File: src/Api/V1/Engine/Api.php [hunk 5]
```diff
@@ -509,10 +509,29 @@ private static function outputXML($statusCode, array $data = null)
...
+        header('HTTP/1.1 ' . self::getDecentHeaderMessage($statusCode));
+        header('content-type: text/xml;charset=' . $charset);
 
         // output XML
         self::$content = $XML->saveXML();
     }
+
+    private static function getDecentHeaderMessage($statusCode)
+    {
+        $possibleCodes = array(
```
Comment (WouterSioen) [inline src/Api/V1/Engine/Api.php]: It seems more logical to me to use this array: https://github.com/symfony/symfony/blob/a44ceb3156e549ffc17baaab7a206be39b636d4f/src/Symfony/Component/HttpFoundation/Response.php#L125. It's more complete and we already have this code loaded trough composer. No need to reinvent the weel :)

File: src/Api/V1/Engine/Api.php [hunk 6]
```diff
@@ -509,10 +509,29 @@ private static function outputXML($statusCode, array $data = null)
...
-        \SpoonHTTP::setHeadersByCode($statusCode);
-        \SpoonHTTP::setHeaders('content-type: text/xml;charset=' . $charset);
+        header('HTTP/1.1 ' . self::getDecentHeaderMessage($statusCode));
+        header('content-type: text/xml;charset=' . $charset);
 
         // output XML
         self::$content = $XML->saveXML();
     }
+
+    private static function getDecentHeaderMessage($statusCode)
```
Comment (WouterSioen) [inline src/Api/V1/Engine/Api.php]: I'd just call this "getHeaderMessage"

File: src/Backend/Core/Engine/Csv.php [hunk 7]
```diff
@@ -45,13 +45,10 @@ public static function outputCSV($filename, array $array, array $columns = null,
...
         $charset = BackendModel::getContainer()->getParameter('kernel.charset');
-        $headers[] = 'Content-type: application/csv; charset=' . $charset;
-        $headers[] = 'Content-Disposition: attachment; filename="' . $filename;
-        $headers[] = 'Content-Length: ' . strlen($csv);
-        $headers[] = 'Pragma: no-cache';
-
-        // overwrite the headers
-        \SpoonHTTP::setHeaders($headers);
+        header('Content-type: application/csv; charset=' . $charset);
+        header('Content-Disposition: attachment; filename="' . $filename);
```
Comment (WouterSioen) [inline src/Backend/Core/Engine/Csv.php]: you miss a quote `"` after your filename

File: src/Backend/Modules/Mailmotor/Config.php [hunk 8]
```diff
@@ -156,7 +174,12 @@ private function checkForGroups()
         // check if there are external groups present in CampaignMonitor
         if ($this->checkForExternalGroups()) {
             // external groups were found, so redirect to the import_groups action
-            \SpoonHTTP::redirect(BackendModel::createURLForAction('ImportGroups', 'Mailmotor'));
+            throw new RedirectException(
```
Comment (WouterSioen) [inline src/Backend/Modules/Mailmotor/Config.php]: A config file should be able to use `$this->redirect` IMO. I'd create a redirect method in the BaseConfig file, that calls this code. It's pretty common to redirect from a config file, so I would hide this Exception hack for the developers and just add a clean $this->redirect.

File: src/Frontend/Modules/Search/Ajax/Livesuggest.php [hunk 9]
```diff
@@ -200,8 +202,11 @@ private function getRealData()
 
         // redirect if the request page doesn't exist
         if ($this->requestedPage > $this->pagination['num_pages'] || $this->requestedPage < 1) {
-            \SpoonHTTP::redirect(
-                FrontendNavigation::getURL(404)
+            throw new RedirectException(
```
Comment (WouterSioen) [inline src/Frontend/Modules/Search/Ajax/Livesuggest.php]: An ajax call that does a redirect. This seems rather strange to me? Shouldn't this just return a response with a 404 statuscode to the JavaScript code calling this method?

File: src/Frontend/Modules/FormBuilder/Widgets/Form.php (hunk 4 above)
Comment (carakas) [inline src/Frontend/Modules/FormBuilder/Widgets/Form.php]: Is there a reason for ?
I think it can be handy to also have a redirect function available in the widgets

File: src/Api/V1/Engine/Api.php (hunk 5 above)
Comment (tijsverkoyen) [inline src/Api/V1/Engine/Api.php]: fixed in 45232a6ae52280028016eb7b570a22770773113b

File: src/Api/V1/Engine/Api.php (hunk 6 above)
Comment (tijsverkoyen) [inline src/Api/V1/Engine/Api.php]: fixed in f3066129a1f957e10bb25895e5f2d58011ed9dd7

File: src/Backend/Core/Engine/Csv.php (hunk 7 above)
Comment (tijsverkoyen) [inline src/Backend/Core/Engine/Csv.php]: fixed in f3066129a1f957e10bb25895e5f2d58011ed9dd7

File: src/Frontend/Modules/Search/Ajax/Livesuggest.php (hunk 9 above)
Comment (tijsverkoyen) [inline src/Frontend/Modules/Search/Ajax/Livesuggest.php]: Correct, but this shouldn't be part of this PR. I created a separate issue: https://github.com/forkcms/forkcms/issues/1256

File: src/Backend/Modules/Mailmotor/Config.php (hunk 8 above)
Comment (tijsverkoyen) [inline src/Backend/Modules/Mailmotor/Config.php]: fixed in 99764ec7b52da378f71a50c96089df1d5a8a3c39

File: src/Api/V1/Engine/Api.php [hunk 10]
```diff
@@ -509,10 +509,19 @@ private static function outputXML($statusCode, array $data = null)
...
-        \SpoonHTTP::setHeadersByCode($statusCode);
-        \SpoonHTTP::setHeaders('content-type: text/xml;charset=' . $charset);
+        header('HTTP/1.1 ' . self::getHeaderMessage($statusCode));
+        header('content-type: text/xml;charset=' . $charset);
 
         // output XML
         self::$content = $XML->saveXML();
     }
+
+    private static function getHeaderMessage($statusCode)
```
Comment (WouterSioen) [inline src/Api/V1/Engine/Api.php]: I'd still add phpdocs. (And even maybe some unit tests).

File: src/Backend/Modules/Error/Actions/Index.php [hunk 11]
```diff
@@ -45,11 +45,11 @@ protected function parse()
         switch ($errorType) {
             case 'module-not-allowed':
             case 'action-not-allowed':
-                \SpoonHTTP::setHeadersByCode(403);
+                header('HTTP/1.1 403 Forbidden');
```
Comment (WouterSioen) [inline src/Backend/Modules/Error/Actions/Index.php]: The getHeaderMessage could be used here?

**Comment by WouterSioen** [comment]
Merged in 1a30797651a25552ee1b738c932a230c68161a39
